libmamba/src/core/download.cpp

```diff
@@ -0,0 +1,609 @@
+#include "mamba/core/download.hpp"
+#include "mamba/core/url.hpp"
+#include "mamba/core/util.hpp"
+#include "mamba/core/util_string.hpp"
+#include "mamba/util/iterator.hpp"
+
+#include "curl.hpp"
+#include "download_impl.hpp"
+
+
+namespace mamba
+{
+
+    /**********************************
+     * DownloadAttempt implementation *
+     **********************************/
+
+    DownloadAttempt::DownloadAttempt(CURLHandle& handle, const DownloadRequest* request)
+        : p_handle(&handle)
+        , p_request(request)
+    {
+        p_stream = make_compression_stream(
+            p_request->url,
+            [this](char* in, size_t size) { return this->write_data(in, size); }
+        );
+    }
+        
+    void DownloadAttempt::prepare_download(CURLMultiHandle& downloader,
+                                           const Context& context,
+                                           on_success_callback success,
+                                           on_failure_callback error)
+    {
+        downloader.add_handle(*p_handle);
+        configure_handle(context);
+        m_success_callback = std::move(success);
+        m_failure_callback = std::move(error);
+        downloader.add_handle(*p_handle);
+    }
+
+    namespace
+    {
+        bool can_retry_download(int http_status, const std::string& url)
+        {
+            return (http_status == 413 || http_status == 429 || http_status >= 500)
```

Comment:
It would be more readable with named constants (so that we get the meaning of these numbers without having to google)